btcutil/tx.go

```diff
@@ -37,30 +38,69 @@ func (t *Tx) MsgTx() *wire.MsgTx {
 
 // Hash returns the hash of the transaction.  This is equivalent to
 // calling TxHash on the underlying wire.MsgTx, however it caches the
-// result so subsequent calls are more efficient.
+// result so subsequent calls are more efficient.  If the Tx has the
+// raw bytes of the tx cached, it will use that and skip serialization.
 func (t *Tx) Hash() *chainhash.Hash {
 	// Return the cached hash if it has already been generated.
 	if t.txHash != nil {
 		return t.txHash
 	}
 
 	// Cache the hash and return it.
-	hash := t.msgTx.TxHash()
+	var hash chainhash.Hash
+
+	// If we have the raw bytes, then don't call msgTx.TxHash as that has the
+	// overhead of serialization.
+	if len(t.rawBytes) > 0 {
+		// If the raw bytes contain the witness, we must strip it out before
+		// calculating the hash.
+		if t.HasWitness() {
+			baseSize := t.msgTx.SerializeSizeStripped()
```

Comment:
Hm, is this same code used when parsing the block in the first place? Feels like one maybe should cache the bytes at that point, to avoid duplicating this parsing logic.

Reply:
https://github.com/btcsuite/btcd/pull/1376 does the caching in-line and at the very lowest level. The diff is also a bit more minimal as well. I think the test case there is just due to a lack of clone before working on a transaction. I've synced with it and haven't encountered that issue in particular.

Reply:
Also the validation logic uses a direct `wire.MsgTx` rather than a `btcutil.Tx`, which is why the memoization above doesn't help us in this case.